### PR TITLE
Feature/s3 upload

### PR DESCRIPTION
**Pull request by Borna-f**
Uploading to s3 works fine. Items remaining:

- Unit tests

### CODE REVIEW

File: examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        )
+
+
+        observer.setTransferListener(object : TransferListener {
+            override fun onStateChanged(id: Int, state: TransferState) {
+                if (state == TransferState.COMPLETED) {
+                    onResponseReceived(ServerResponse.successfulEmpty())
+                    Log.i("UtilityObserver","Upload Finished!");
+                } else if (state == TransferState.FAILED) {
+                    Log.i("UtilityObserver","Upload Failed!");
```
Comment (barchard) [inline examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt]: why is this not handled and propagated up?

File: examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                } else {
+                    Log.i("UtilityObserver","Unkown!");
+                }
+            }
+
+            override fun onProgressChanged(id: Int, bytesCurrent: Long, bytesTotal: Long) {
+                resetUploadedBytes()
```
Comment (barchard) [inline examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt]: why is this called? this sounds like it will cause the UI to jump around (jump to 0 then back up). seems like you just need to overwrite the progress values instead of resetting them each time.

File: examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                totalBytes = bytesTotal
+                onProgress(bytesCurrent)
+                val progress = bytesCurrent.toDouble() / bytesTotal * 100
+                Log.i("UtilityObserver","Progress: " + progress);
+            }
+
+            override fun onError(id: Int, ex: Exception) {
+                Log.e("UtilityObserver","Error Occured");
```
Comment (barchard) [inline examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt]: why is this not recorded? why is the reason of error not logged?

File: examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import com.amazonaws.services.s3.AmazonS3Client
+import com.amazonaws.services.s3.model.CannedAccessControlList
+import net.gotev.uploadservice.UploadTask
+import net.gotev.uploadservice.network.HttpStack
+import net.gotev.uploadservice.network.ServerResponse
+import java.io.File
+
+class S3UploadTask : UploadTask() {
+
+    private val s3params by lazy {
```
Comment (barchard) [inline examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt]: S in `S3` is typically capitalized, especially for class names

**Review by barchard — COMMENTED** [review]
@himanshuagarwal456 thoughts here. also, this should go into `develop`, not master

File: examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadRequest.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package com.example.uploadservice.s3
```
Comment (himanshuagarwal456) [inline examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadRequest.kt]: suggest using same path as ftp module
uploadservice-ftp/src/main/java/net/gotev/uploadservice/ftp.
instead of com/example

File: examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        s3.getUrl(s3params.bucket_name, s3params.serverSubpath)
+        val transferUtility = TransferUtility.builder().s3Client(s3).context(context).build()
+        val file = File(s3params.uploadFilepath)
+        require(file.exists()) { "Error! Please choose a valid file for upload" }
+        TransferNetworkLossHandler.getInstance(context)
+        val observer = transferUtility.upload(
+                s3params.bucket_name,
+                s3params.serverSubpath,
+                file,
+                CannedAccessControlList.Private
```
Comment (himanshuagarwal456) [inline examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt]: This aws stuff could/should be moved into its own class similar to ftpclientwrapper.

File: examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTaskParameters.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import kotlinx.android.parcel.Parcelize
+import net.gotev.uploadservice.persistence.Persistable
+import net.gotev.uploadservice.persistence.PersistableData
+import java.io.File
+
+@Parcelize
+data class S3UploadTaskParameters (
+        var uploadFilepath : String,
+        var serverSubpath: String = "",
+        var bucket_name: String,
```
Comment (himanshuagarwal456) [inline examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTaskParameters.kt]: bucket_name - > bucketName. Same for other variables which are using snake case, change to camelcase.

File: examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt [hunk 8]
```diff
@@ -0,0 +1,66 @@
+package com.example.uploadservice.s3
```
Comment (himanshuagarwal456) [inline examples/app/uploadservice-s3/src/main/java/com/example/uploadservice/s3/S3UploadTask.kt]: I think entire s3 module is duplicated under two different paths 
android-upload-service/uploadservice-s3/ and android-upload-service/examples/app/uploadservice-s3. We only need the former.

File: examples/app/.idea/gradle.xml [hunk 9]
```diff
@@ -15,6 +15,7 @@
             <option value="$PROJECT_DIR$/../../uploadservice" />
             <option value="$PROJECT_DIR$/../../uploadservice-ftp" />
             <option value="$PROJECT_DIR$/../../uploadservice-okhttp" />
+            <option value="$PROJECT_DIR$/../../uploadservice-s3" />
```
Comment (himanshuagarwal456) [inline examples/app/.idea/gradle.xml:18]: 👍

**Review by himanshuagarwal456 — APPROVED** [review]

File: uploadservice/src/main/java/net/gotev/uploadservice/UploadTask.kt [hunk 10]
```diff
@@ -156,6 +159,11 @@ abstract class UploadTask : Runnable {
         }
     }
 
+    /**
+     * If an exception is thrown while uploading this method will be executed to do the retry.
+     * If you want you require to call this method directly on error, you can do so on your upload task
+     * See S3 implementation for example
+     */
```
Comment (barchard) [inline uploadservice/src/main/java/net/gotev/uploadservice/UploadTask.kt:177]: you need to add `@param (type) Description` for doxygen style comments

File: uploadservice/src/main/java/net/gotev/uploadservice/UploadTask.kt [hunk 11]
```diff
@@ -128,6 +128,9 @@ abstract class UploadTask : Runnable {
         errorDelay = UploadServiceConfig.retryPolicy.initialWaitTimeSeconds.toLong()
     }
 
+    /**
+     * This and runUpload will run the upload task
```
Comment (barchard) [inline uploadservice/src/main/java/net/gotev/uploadservice/UploadTask.kt]: This needs to be a better description.

>Executes the upload in a background thread and notifies each observer of the upload progress.

Soemthing like that (if true).

**Review by barchard — CHANGES_REQUESTED** [review]

File: uploadservice-s3/src/main/java/net/gotev/uploadservice/s3/S3UploadTask.kt [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }
+
+    /**
+     * Calculates the total bytes of this upload task.
+     * This the sum of all the lengths of the successfully uploaded files and also the pending
+     * ones.
+     */
+    private fun calculateUploadedAndTotalBytes() {
+        resetUploadedBytes()
```
Comment (barchard) [inline uploadservice-s3/src/main/java/net/gotev/uploadservice/s3/S3UploadTask.kt:55]: what is this, where is it defined and what does it do? is it required? this seems like it will cause the UI to jump around. Why is this necessary? It seems like all you need to do is calculate the total bytes and call onProgress

File: uploadservice-s3/src/main/java/net/gotev/uploadservice/s3/S3ClientWrapper.kt [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        UploadServiceLogger.debug(javaClass.simpleName, uploadId) {
+            "Starting S3 upload of: ${uploadFile.handler.name(context)}"
+        }
+        uploadingFile = uploadFile
+        val file = File(uploadFile.path)
+        transferObserver = transferUtility.upload(
+                bucketName,
+                serverSubPath + "/" + file.name,
+                file,
+                CannedAccessControlList.Private
```
Comment (barchard) [inline uploadservice-s3/src/main/java/net/gotev/uploadservice/s3/S3ClientWrapper.kt]: shouldn't the default access be part of the params for the file or upload? we will never change this, but would end users?

**Review by barchard — CHANGES_REQUESTED** [review]

File: uploadservice-s3/src/main/java/net/gotev/uploadservice/s3/S3UploadTask.kt (hunk 12 above)
Comment (Borna-f) [reply uploadservice-s3/src/main/java/net/gotev/uploadservice/s3/S3UploadTask.kt:55]: This is not my code. I am not sure if it works properly if we remove it. I just kept it the same as FTP.

https://github.com/gotev/android-upload-service/blob/ff518f03159c5b1e7b2b001c75d26fbcc22c542a/uploadservice-ftp/src/main/java/net/gotev/uploadservice/ftp/FTPUploadTask.kt#L80

Comment (Borna-f) [reply uploadservice-s3/src/main/java/net/gotev/uploadservice/s3/S3UploadTask.kt:55]: But generally, the uploaded bytes is cumulative: uploadedBytes += bytesSent
if we don't reset it each time and if it has some values previously it makes it buggy. This method basically resets uploaded bytes and sums again how many bytes are sent. There might be a better implementation but that probably requires changing the base UploadTask class.

**Review by barchard — APPROVED** [review]
